### ISSUE TITLE
compille error with flag -DUA_ENABLE_AMALGAMATION=ON

### ISSUE THREAD

**Issue by Kandimus** [open62541/open62541#3695]
Hello. GJ.

I decided to use the flags "**-DUA_ENABLE_AMALGAMATION=ON DUA_ENABLE_ENCRYPTION_OPENSSL=ON**" to compile open62541 project and use one h-file in my project.  This errors occurred during compilation:
```
[ 65%] Building C object CMakeFiles/open62541-object.dir/open62541.c.o
In file included from /mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:28:0:
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.h:31132:41: error: unknown type name ‘X509’
                                         X509 *                publicKeyX509,
                                         ^~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.h:31147:30: error: unknown type name ‘X509’
                              X509 *          publicX509);
                              ^~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.h:31155:37: error: unknown type name ‘X509’
 UA_Openssl_RSA_Public_GetKeyLength (X509 *     publicKeyX509,
                                     ^~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.h:31185:60: error: unknown type name ‘X509’
 UA_OpenSSL_X509_compare (const UA_ByteString * cert, const X509 * b);
                                                            ^~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.h:31191:39: error: unknown type name ‘X509’
                                       X509 *                publicKeyX509,
                                       ^~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.h:31219:35: error: unknown type name ‘X509’
                                   X509 *          publicX509);
                                   ^~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.h:31963:2: error: #endif without #if
 #endif /* OPEN62541_H_ */
  ^~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:58974:1: error: no previous prototype for ‘UA_OpenSSL_RSA_PKCS1_V15_SHA256_Verify’ [-Werror=missing-prototypes]
 UA_OpenSSL_RSA_PKCS1_V15_SHA256_Verify (const UA_ByteString * msg,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:59161:1: error: no previous prototype for ‘UA_Openssl_RSA_OAEP_Encrypt’ [-Werror=missing-prototypes]
 UA_Openssl_RSA_OAEP_Encrypt (UA_ByteString * data,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:59260:1: error: no previous prototype for ‘UA_Openssl_RSA_Public_GetKeyLength’ [-Werror=missing-prototypes]
 UA_Openssl_RSA_Public_GetKeyLength (X509 *     publicKeyX509,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:59533:1: error: conflicting types for ‘UA_OpenSSL_X509_compare’
 UA_OpenSSL_X509_compare (const UA_ByteString * cert,
 ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:28:0:
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.h:31185:1: note: previous declaration of ‘UA_OpenSSL_X509_compare’ was here
 UA_OpenSSL_X509_compare (const UA_ByteString * cert, const X509 * b);
 ^~~~~~~~~~~~~~~~~~~~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:59551:1: error: no previous prototype for ‘UA_OpenSSL_RSA_PKCS1_V15_SHA1_Verify’ [-Werror=missing-prototypes]
 UA_OpenSSL_RSA_PKCS1_V15_SHA1_Verify (const UA_ByteString * msg,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:59688:1: error: no previous prototype for ‘UA_Openssl_RSA_PKCS1_V15_Encrypt’ [-Werror=missing-prototypes]
 UA_Openssl_RSA_PKCS1_V15_Encrypt (UA_ByteString * data,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:60561:1: error: redefinition of ‘UA_Asy_compareCertificateThumbprint’
 UA_Asy_compareCertificateThumbprint (const UA_SecurityPolicy * securityPolicy,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:59988:1: note: previous definition of ‘UA_Asy_compareCertificateThumbprint’ was here
 UA_Asy_compareCertificateThumbprint (const UA_SecurityPolicy * securityPolicy,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:60577:1: error: redefinition of ‘UA_Asy_makeCertificateThumbprint’
 UA_Asy_makeCertificateThumbprint (const UA_SecurityPolicy * securityPolicy,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/hgfs/G/_3tonn/thirdparty.scr/open62541-master/build/open62541.c:60004:1: note: previous definition of ‘UA_Asy_makeCertificateThumbprint’ was here
 UA_Asy_makeCertificateThumbprint (const UA_SecurityPolicy * securityPolicy,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: error: unrecognized command line option ‘-Wno-static-in-inline’ [-Werror]
cc1: all warnings being treated as errors
CMakeFiles/open62541-object.dir/build.make:278: recipe for target 'CMakeFiles/open62541-object.dir/open62541.c.o' failed
make[2]: *** [CMakeFiles/open62541-object.dir/open62541.c.o] Error 1
CMakeFiles/Makefile2:396: recipe for target 'CMakeFiles/open62541-object.dir/all' failed
make[1]: *** [CMakeFiles/open62541-object.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```
When I used the flags "**-DUA_ENABLE_AMALGAMATION=ON DUA_ENABLE_ENCRYPTION_OPENSSL=OFF**" compilation is successful.

As I understand it, openssl headers are not included in the shared file.
OS Debian.

## Checklist

Please provide the following information:

 - [+] open62541 Version (release number or git tag): v1.1
 - [+] Other OPC UA SDKs used (client or server): no
 - [+] Operating system: Dibian
 - [-] Logs (with `UA_LOGLEVEL` set as low as necessary) attached
 - [-] Wireshark network dump attached
 - [-] Self-contained code example attached
 - [-] Critical issue

**Comment by masongk** [comment]
@Kandimus You need to enable the UA_ENABLE_ENCRYPTION=ON to resolve the issue!

**Comment by Kandimus** [comment]
Re: [open62541/open62541] compille error with flag -DUA_ENABLE_AMALGAMATION=ON (#3695)


Здравствуйте, MasonMakrov.

My make file:

cmake \
-DUA_DEBUG=ON \
-DCMAKE_BUILD_TYPE=RelWithDebInfo \
-DUA_LOGLEVEL=300 \
-DUA_MULTITHREADING=100 \
-DUA_ENABLE_SUBSCRIPTIONS=ON \
-DUA_ENABLE_SUBSCRIPTIONS_EVENTS=ON \
-DUA_ENABLE_SUBSCRIPTIONS_ALARMS_CONDITIONS=OFF \
-DUA_ENABLE_METHODCALLS=ON \
-DUA_ENABLE_NODEMANAGEMENT=OFF \
-DUA_BUILD_EXAMPLES=ON \
-DUA_ENABLE_LOG_COLORS=OFF \
-DUA_ENABLE_AMALGAMATION=ON \
-DUA_ENABLE_DISCOVERY=ON \
-DUA_NAMESPACE_ZERO=REDUCED \
-DUA_ENABLE_TYPEDESCRIPTION=ON \
-DUA_ENABLE_STATUSCODE_DESCRIPTIONS=ON \
-DUA_ENABLE_HISTORIZING=ON \
-DUA_ENABLE_ENCRYPTION=ON \
-DUA_ENABLE_ENCRYPTION_OPENSSL=ON \
-DUA_BUILD_SELFSIGNED_CERTIFICATE=ON \
..

make


-- 
С уважением,
Литвинов Виталий                         mailto:vedun@mail.ru

Вы писали 28 июня 2020 г., 20:24:55:





@Kandimus You need to enable the UA_ENABLE_ENCRYPTION=ON to resolve the issue!

**Comment by jpfr** [comment]
Fixed in #3753
